src/corehost/cli/fxr/hostfxr.cpp

```diff
@@ -396,55 +394,452 @@ SHARED_API hostfxr_error_writer_fn hostfxr_set_error_writer(hostfxr_error_writer
     return trace::set_error_writer(error_writer);
 }
 
-coreclr_delegate_type hostfxr_delegate_to_coreclr_delegate(hostfxr_delegate_type type)
+namespace
+{
+    int populate_startup_info(const hostfxr_initialize_parameters *parameters, host_startup_info_t &startup_info)
+    {
+        if (parameters != nullptr)
+        {
+            if (parameters->host_path != nullptr)
+                startup_info.host_path = parameters->host_path;
+
+            if (parameters->dotnet_root != nullptr)
+                startup_info.dotnet_root = parameters->dotnet_root;
+        }
+
+        if (startup_info.host_path.empty())
+        {
+            if (!pal::get_own_executable_path(&startup_info.host_path) || !pal::realpath(&startup_info.host_path))
+            {
+                trace::error(_X("Failed to resolve full path of the current host [%s]"), startup_info.host_path.c_str());
+                return StatusCode::CoreHostCurHostFindFailure;
+            }
+        }
+
+        if (startup_info.dotnet_root.empty())
+        {
+            pal::string_t mod_path;
+            if (!pal::get_own_module_path(&mod_path))
+                return StatusCode::CoreHostCurHostFindFailure;
+
+            pal::string_t fxr_root = get_directory(get_directory(mod_path));
```

Comment:
In your new PR #6107 the logic to find the dotnet folder is slightly diffrerent: https://github.com/dotnet/core-setup/blob/5febb4e61d11b492dc0f7183747a489840afaa6a/src/corehost/cli/fxr_resolver.h#L36
They should be the same, ideally share the same function to do this.

Reply:
Rebased and used the same function (moved it into utils, since it seemed weird to have hostfxr pull in fxr_resolver)